frontend/src/components/main-page/groups-tree-page/service-details/dependencies/visualization/visualization-modal.tsx

```diff
@@ -0,0 +1,228 @@
+                data={controller.sankeyData}
+                colors={(node): string => node.customData.color}
+                label={(node): string => node.customData.label}
+                linkTooltip={CustomLinkTooltip}
+                align={(node, maxDepth): number =>
+                  getNodeAlignment(node as SankeyNodeWithDepth, maxDepth)
+                }
+                enableLinkGradient
+              />
+            </Box>
+          )}
+        </Box>
+      </Box>
+    </Dialog>
+  );
+};
+
+interface State {
+  sankeyConfig: SankeyConfig;
+}
+interface Controller {
+  state: State;
+
+  sankeyData: SankeyData;
+
+  setSankeyConfig: (newConfig: SankeyConfig) => void;
+}
+function useController(props: Props): Controller {
+  const serviceDocsService = useServiceDocsServiceContext();
+
+  const [state, setState] = React.useState<State>({
+    sankeyConfig: {
+      pivotService: props.pivotService,
+
+      includeAPIs: true,
+      includeEvents: true,
+    },
+  });
+
+  const sankeyData = React.useMemo(
+    () =>
+      convertServiceDocsToSankeyData(
+        serviceDocsService.serviceDocs,
+        state.sankeyConfig,
+      ),
+    [serviceDocsService.serviceDocs, state.sankeyConfig],
+  );
+
+  return {
+    state: state,
+
+    sankeyData: sankeyData,
+
+    setSankeyConfig: (newConfig): void => {
+      setState((state) => ({ ...state, sankeyConfig: newConfig }));
+    },
+  };
+}
+
+interface SankeyNodeWithDepth extends CustomSankeyNode {
+  depth: number;
+}
+function getNodeAlignment(node: SankeyNodeWithDepth, maxDepth: number): number {
+  // Important: It is possible that the diagram has less than 3 columns (e.g. if there are no source nodes). This makes it hard to determine where a particular node shall be positioned. To solve this issue for now, we simply use the node depth. This is probably not ideal (it could produce unexpected alignments). However, this should be fine for now given that most of our diagrams contain enough data to produce 3-column layouts.
+
+  if (maxDepth < 2) {
+    return node.depth;
+  }
+
+  switch (node.customData.position) {
+    case NodePosition.Left:
+      return 0;
+    case NodePosition.Middle:
+      return 1;
+    case NodePosition.Right:
+      return 2;
+  }
+}
+
+/**
+ * The tooltip that is being shown when hovering a link.
+ */
+const CustomLinkTooltip: React.FC<{
+  link: SankeyLinkDatum<CustomSankeyNode, CustomSankeyLink>;
+}> = (props) => {
+  return (
+    <Paper
+      elevation={1}
+      sx={{
+        background: '#ffffff',
+        paddingX: 2,
+        paddingY: 1,
+        display: 'flex',
+        gap: 2,
+        alignItems: 'center',
+      }}
+    >
+      <Box
+        sx={{
+          background: props.link.source.color,
+          height: '15px',
+          width: '15px',
+        }}
+      />
+
+      <Box sx={{ display: 'flex', alignItems: 'center', gap: 2 }}>
+        <span>{props.link.source.label}</span>
+        <Icons.ArrowForward />
```

Comment:
Can we have different tooltips depending on if event or API?
I could imagine that showing the following would be valuable to users:
* "WebClient consumes API /notifications/configs"
* "NotificationService provides API /notifications/configs"
* "LoadService produces event load.execution.success"
* "NotificationService consumes event load.execution.success"

If we decide to implement something like this it should be a separate issue.

Reply:
Yes, that should be possible 👍